### PR TITLE
Improve CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by etpinard**
... and resolves https://github.com/plotly/plotly.js/issues/66 with https://github.com/plotly/plotly.js/commit/278ed7cf71304907cef2c4593e4430d82760f685

cc @plotly/plotly_js please let me know what you think!

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -248,8 +248,8 @@ All traces modules defined:
   This object is used to generate the plot-schema JSON.
 - `_module.supplyDefaults`: Takes in input trace settings and coerces them into "full" settings
   under `gd._fullData`. This one is called during the figure-wide `Plots.supplyDefaults` routine.
-  Note that the `suppyDefaults` method performance should scale with the number of attribute **not** the
-  number of data points.
+  Note that the `suppyDefaults` method performance should scale with the number of attributes (**not** the
+  number of data points).
```
Comment (alexcjohnson) [inline CONTRIBUTING.md]:
```suggestion
  Note that the `supplyDefaults` method performance should scale with the number of attributes (**not** the
  number of data points - so it should not loop over any data arrays).
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -258,13 +258,13 @@ All traces modules defined:
 
 Other methods used by some trace modules:
 
-- `_module.categories`: list of string identifies used to grouped traces by behavior
+- `_module.categories`: list of string identifiers used to group traces by behavior
```
Comment (alexcjohnson) [inline CONTRIBUTING.md]: Can you say something about what these categories are used for, and how to figure out what categories to use for your new trace?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -273,8 +273,7 @@ Other methods used by some trace modules:
   selections.
 - `_module.convert`: Sometimes separated from `_module.plot` or `_module.calc` to convert the
   plotly.js settings to another framework e.g. to `gl-plot3d` for `gl3d` traces, to
-  `mapbox-gl` from `mapbox` traces. This split can also make the logic easier to
-  test.
+  `mapbox-gl` from `mapbox` traces. This split can make the logic easier to test.
```
Comment (alexcjohnson) [inline CONTRIBUTING.md]: What pathway calls `convert`? ie is it called by `plot`, by `calc`, or from a higher level of the framework?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -196,18 +216,64 @@ This will produce the following plot, and say you want to simulate a selection p
...
+
+The trace module methods are meant to be called as part of loops during subplot-specific
+(e.g. in `plots/cartesian/index.js`) and figure-wide (e.g. in `plots/plots.js`) subroutines.
+That way, the subroutines work no matter which trace modules got registered.
+
+All traces modules set:
+
+- `_module.name`: name of the trace module as used by the trace `type` attribute.
+- `_module.basePlotModule`: base plot (or subplot) module corresponding to the
+  trace type (e.g. `scatter` links up the `Cartesian` base plot module, `scatter3d` links up `gl3d`).
```
Comment (alexcjohnson) [inline CONTRIBUTING.md]:
```suggestion
  trace type (e.g. `scatter` links to the `Cartesian` base plot module, `scatter3d` links to `gl3d`).
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -196,18 +216,64 @@ This will produce the following plot, and say you want to simulate a selection p
...
+- `_module.attributes`: JSON-serializable object of attribute declarations.
+  This object is used to generate the plot-schema JSON.
+- `_module.supplyDefaults`: Takes in input trace settings and coerces them into "full" settings
+  under `gd._fullData`. This one is called during the figure-wide `Plots.supplyDefaults` routine.
+  Note that the `suppyDefaults` method performance should scale with the number of attributes (**not** the
+  number of data points).
+- `_module.calc`: Converts inputs data into "calculated" (or sanitized) data. This one is called during
+  the figure-wide `Plots.doCalcdata` routine. The `calc` method is allowed to
+  scale with the number of data points and is in general more costly than `supplyDefaults`.
+  Please note that some edit pathways skip `Plots.doCalcdata`.
```
Comment (alexcjohnson) [inline CONTRIBUTING.md]:
```suggestion
  Please note that some edit pathways skip `Plots.doCalcdata`. This is determined by the
  `editType` flags in the attributes files.
```

File: CONTRIBUTING.md (hunk 3 above)
Comment (etpinard) [reply CONTRIBUTING.md]: during `plot`:

- all `gl3d` traces
- all `mapbox` traces
- `heatmapgl`, `contourgl`, `pointcloud`

during `calc`:
- `scattergl`, `scatterpolargl`

Comment (alexcjohnson) [reply CONTRIBUTING.md]: OK good - the key for contributors is to know it's not hooking into a predefined pathway in the framework.
```suggestion
  `mapbox-gl` from `mapbox` traces. This split can make the logic easier to test.
  If you make a `convert`, you should call it from either `calc` or `plot`.
```

**Pull request merged**